src/pyscheduling/SMSP/SingleMachine.py

```diff
@@ -440,485 +448,119 @@ def toJSON(self):
-            for i in range(startIndex, job_schedule_len):
-                job_i = self.job_schedule[i].id
-
-                if hasattr(instance, 'R'):
-                    startTime = max(ci, instance.R[job_i])
-                else:
-                    startTime = ci
-                if hasattr(instance, 'S'):
-                    setupTime = instance.S[job_prev_i][job_i]
-                else:
-                    setupTime = 0
-                proc_time = instance.P[job_i]
-                ci = startTime + setupTime + proc_time
-
-                self.job_schedule[i] = Job(job_i, startTime, ci)
-                wiTi += instance.W[job_i]*max(ci-instance.D[job_i], 0)
-                self.wiTi_index[i] = wiTi
-                job_prev_i = job_i
-        self.objective = wiTi
-        return wiTi
-
-    def total_weighted_lateness_insert(self, job: int, pos: int, instance: SingleInstance):
-        """Computes the machine's total wighted lateness if we insert "job" at "pos" in the machine's job_schedule
-        Args:
-            job (int): id of the inserted job
-            pos (int): position where the job is inserted in the machine
-            instance (SingleInstance): the current problem instance
-        Returns:
-            int : total weighted lateness
-        """
-        if pos > 0:
-            c_prev = self.job_schedule[pos - 1].end_time
-            job_prev_i = self.job_schedule[pos - 1].id
-            if hasattr(instance, 'R'):
-                release_time = max(instance.R[job] - c_prev, 0)
-            else:
-                release_time = 0
-            if hasattr(instance, 'S'):
-                setupTime = instance.S[job_prev_i][job]
-            else:
-                setupTime = 0
-            ci = c_prev + release_time + setupTime + instance.P[job]
-            wiTi = self.wiTi_index[pos - 1]+instance.W[job]*ci
+        if startIndex > 0:
+            ci = self.job_schedule[startIndex - 1].end_time
+            obj = obj_cache[startIndex - 1] if obj_cache is not None else 0 # When the cache is not specified, the objective does not depend on it (like Cmax)
+            job_prev_i = self.job_schedule[startIndex - 1].id
         else:
-            ci = instance.S[job][job] + instance.P[job]
-            wiTi = instance.W[job]*ci
-        job_prev_i = job
-        for i in range(pos, len(self.job_schedule)):
-            job_i = self.job_schedule[i][0]
-
-            if hasattr(instance, 'R'):
-                startTime = max(ci, instance.R[job_i])
-            else:
-                startTime = ci
-            if hasattr(instance, 'S'):
-                setupTime = instance.S[job_prev_i][job_i]
-            else:
-                setupTime = 0
-            proc_time = instance.P[job_i]
-            ci = startTime + setupTime + proc_time
-            wiTi += instance.W[job_i]*max(ci-instance.D[job_i], 0)
-
-            job_prev_i = job_i
+            ci, obj, job_prev_i,  = 0, 0, self.job_schedule[startIndex].id
```

Comment:
When we assign the value of self.job_schedule[startIndex].id to job_prev_i, this means that the job_schedule is not empty (contains one element), why in this case ci and obj are equal to 0 ?

Reply:
In this case, startIndex = 0. Therefore the job_prev_i is the first job in the schedule. 
This is to allow computing the first setup S[i][i]. And the previous ci and obj are just starting so they're equal to 0 (there's no job before startIndex).